// managercms-sdk/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ManagerCMS = exports.ManagerCMSError = void 0;
class ManagerCMSError extends Error {
    status;
    data;
    constructor(status, message, data) {
        super(message);
        this.status = status;
        this.data = data;
        this.name = 'ManagerCMSError';
    }
}
exports.ManagerCMSError = ManagerCMSError;
class ManagerCMS {
    apiUrl;
    token;
    constructor(token, apiUrl = 'https://manager.1bits.site') {
        this.apiUrl = apiUrl.replace(/\/$/, '');
        this.token = token;
    }
    async request(endpoint, options = {}) {
        const response = await fetch(`${this.apiUrl}${endpoint}`, {
            ...options,
            headers: {
                'Authorization': `Bearer ${this.token}`,
                'Content-Type': 'application/json',
                ...options.headers,
            }
        });
        if (!response.ok) {
            let errorData;
            try {
                errorData = await response.json();
            }
            catch {
                errorData = null;
            }
            throw new ManagerCMSError(response.status, `Error en la API de ManagerCMS: ${response.statusText}`, errorData);
        }
        return response.json();
    }
    // --- Endpoints de Delivery ---
    async getWebsiteInfo() {
        return this.request('/websites/');
    }
    async getContentTypes() {
        return this.request('/websites/content-types/');
    }
    async getEntries(modelSlug, options = {}) {
        const params = new URLSearchParams();
        if (options.pageSize)
            params.append('page_size', options.pageSize.toString());
        if (options.page)
            params.append('page', options.page.toString());
        if (options.ordering)
            params.append('ordering', options.ordering);
        if (options.search)
            params.append('search', options.search);
        if (options.filters) {
            Object.entries(options.filters).forEach(([key, value]) => {
                params.append(key, value.toString());
            });
        }
        const queryString = params.toString();
        const baseUrl = options.pageSize || options.page
            ? `/websites/paginacion/content/${modelSlug}/entries/`
            : `/websites/content/${modelSlug}/entries/`;
        const url = queryString ? `${baseUrl}?${queryString}` : baseUrl;
        return this.request(url);
    }
    async getEntry(modelSlug, id) {
        return this.request(`/websites/content/${modelSlug}/entries/${id}/`);
    }
    // --- Management API ---
    async createEntry(modelSlug, data) {
        return this.request(`/websites/content/${modelSlug}/entries/`, {
            method: 'POST',
            body: JSON.stringify(data),
        });
    }
    async updateEntry(modelSlug, id, data) {
        return this.request(`/websites/content/${modelSlug}/entries/${id}/`, {
            method: 'PUT',
            body: JSON.stringify(data),
        });
    }
    async deleteEntry(modelSlug, id) {
        return this.request(`/websites/content/${modelSlug}/entries/${id}/`, {
            method: 'DELETE',
        });
    }
}
exports.ManagerCMS = ManagerCMS;
